### DISCUSSION TITLE
Nano RP2040 Connect temp reading never changes

### DISCUSSION THREAD

**Discussion by pwl1821** [micropython/micropython#9454] [Using MicroPython] [answered]
Hi,

I have a Arduino Nano RP2040 Connect utlising a 10K thermistor & 10K resistor. Its input is on A0. The MicroPython code I'm using continously displays 24C (varying only +/- 0.02C) in Thonny's serial monitor. It does this even when I completely disconnect the thermistor.
(As a test, I'm able to get good temperature readings with an Arduino sketch on an identical ciruit on the same board.)

My programmer is conversant in Python, but new to MicroPython, and microcontrollers.
We're a bit baffled. Any suggestions of what's wrong?

Code:
```
from machine import ADC, Pin
import math
import time

adc = ADC(Pin(26))  # create ADC object on ADC pin

RT0 = 10000
B = 3950
VCC = 3.3
R = 10000
T0 = 25 + 273.15


while True:
    # read value, 0-65535 across voltage range 0.0v - 3.3v. equivalent to - VRT = analogRead(A0)
    VRT = adc.read_u16()
    VRT = (3.3 / 1023.00) * VRT  # Conversion to voltage
    VR = VCC - VRT
    RT = VRT / (VR / R) * -1
    ln = math.log(RT / RT0)
    TX = (1 / ((ln / B) + (1 / T0)))  # Temperature from thermistor
    TX = TX - 273.15  # Conversion to Celsius
    print("Temperature: ", TX)    
    time.sleep_ms(500)  # sleep for 500 milliseconds
```
Circuit:
![image](https://user-images.githubusercontent.com/4216512/192886494-f2a3f74d-7e82-415b-a66a-3d5fe3fc130a.png)

**Comment by davefes** [comment]
First thing I would do is:
```py
print (str(VT))
```
and touch the thermistor.

Is pin 26 allowed for an ADC?  On the ESP32 you can't use ADC2 ... maybe there are some restrictions on your port.

**Reply by pwl1821** [reply]
Pin 26 is A0 on the ANC, so that worked.

**Comment by scruss** [comment, marked as answer]
The line
```
    VRT = (3.3 / 1023.00) * VRT  # Conversion to voltage
```
should probably be
```
    VRT = (3.3 / 65535) * VRT  # Conversion to voltage
```
since read_u16() returns 0 .. 65535.

I have known-good thermistor code [here](https://forums.raspberrypi.com/viewtopic.php?t=336542#p2014365) that will need a little adaptation to your use case. I really should finish writing that generic thermistor module ...

**Reply by pwl1821** [reply]
> I have known-good thermistor code [here](https://forums.raspberrypi.com/viewtopic.php?t=336542#p2014365)

Your code worked well with a couple of minor edits below.
#led = Pin(led_builtin, Pin.OUT)          # built in LED
adc = ADC(26)                    # PL edit: ANC GPIO 26  / physical pin #?

The readings were for the most part quite stable – only changing 0.1C and occasional jumping of 0.2-0.3C.

Near perfect accuracy at room temp compared to two digital thermometers - +/- 0.0C to +0.2C.
In the kitchen oven set at 50C, accuracy was - +/ 0.2C to 0.3C. In the fridge it varied between - +/- 0.2C to 0.6C. Most often reading a bit high, rather than low.

My next step will be to get the Steinhart-Hart calculation sorted out. I need +/- 0.3 to +/-0.6C max, through the 0-100C range.

Thanks heaps.

**Reply by scruss** [reply]
You're going to need a much more stable voltage reference than the Pico/Nano RP2040 uses and a precision thermistor (aka not Adafruit's "precision" one) with 3 or 4 constants specified: Vishay have ones like this. Similarly, a 1% or better thermally-stable resistor on the high side. The RP2040's ADC has known issues, and very conservative types only trust it to 8 bits of precision.

The Vishay NTC datasheet/spreadsheet (only works on real Excel, alas) has all the calculations and descriptions. There are thermistor modules for Python (not MicroPython) that implement the full-form Steinhart-Hart equation properly. The RepRap and Marlin open source 3d printer projects have a wealth of information on how to use thermistors well, though much of their code presumes you're using a 5 V Arduino.

**Comment by robert-hh** [comment]
> . It does this even when I completely disconnect the thermistor.

That looks like a wiring error. Lacking a thermistor, I made the setup using an LDR. That shows varying values by changing the brightness and a slight change of your code:
```

from machine import ADC, Pin
import math
import time

adc = ADC(Pin(26))  # create ADC object on ADC pin

RT0 = 10000
B = 3950
VCC = 3.3
R = 10000
T0 = 25 + 273.15


while True:
    # read value, 0-65535 across voltage range 0.0v - 3.3v. equivalent to - VRT = analogRead(A0)
    VRT = adc.read_u16()
    VRT = (3.3 / 65535.0) * VRT  # Conversion to voltage ### changed here
    VR = VCC - VRT
    RT = VRT / (VR / R)  ### changed here, otherwise there was a domain error for log.
    ln = math.log(RT / RT0)
    TX = (1 / ((ln / B) + (1 / T0)))  # Temperature from thermistor
    TX = TX - 273.15  # Conversion to Celsius
    print("Temperature: ", TX)
    time.sleep_ms(500)  # sleep for 500 milliseconds
```

**Reply by pwl1821** [reply]
Your code also worked quite well with a couple of minor edits below.

Having the   `  time.sleep_ms(500)  # sleep for 500 milliseconds` line was helpful. I changed it to (2000) to slow down the reads.

Like @[scruss ](https://github.com/scruss) the readings were pretty stable – only changing 0.1 – 0.3C and occasional brief jumping of 0.2-0.7C.

Again, near perfect accuracy at room temp compared to two digital thermometers - +/- 0.0C to +0.2C. In the kitchen oven set at 50C, accuracy was - +/ 0.2C to 0.7C. In the fridge it varied between - +/- 0.2C to 0.5C.

My wiring seems to be working sufficiently for these tests.

Great trick using the LDR. Thanks much.

**Reply by pwl1821** [reply]
BTW - Would have also marked yours as an answer, but can't figure out how to mark two different posts as answers.

**Comment by steveh127** [comment]
The first thing I would do is print the raw reading from the ADC. You should then be able to see if the problem is with the readings not varying or an issue with the conversion code.

**Reply by pwl1821** [reply]
Don't know how to do that but would be good knowledge. What would the code be?

**Reply by scruss** [reply]
`print(adc.read_u16())`
